File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.java
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rLen = matrix.length, cLen = matrix[0].length;
        int low = 0, high = rLen*cLen-1;
        int mid;
        while(low<=high) {
            mid = low + (high-low)/2;

            int lowN = matrix[low/cLen][low%cLen];
            int midN = matrix[mid/cLen][mid%cLen];

            /* target 값이면 그대로 반환 */
            if(midN == target) return true;
            
            if(midN >= target && lowN <= target) high = mid-1;
            else low = mid+1;
        }
        
        return false;
    }
}

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        m, n = len(matrix), len(matrix[0])
        
        mid = m * n // 2
        l, r = 0, m * n
        while l < r:
            row, col = mid//n, mid%n-1
            mid_val = matrix[row][col]

            if mid_val == target:
                return True
            if mid_val < target:
                l = mid + 1
            else:
                r = mid - 1
            mid = (r + l) // 2

        return False